src/stacked_borrows.rs

```diff
@@ -467,6 +469,91 @@ impl<'tcx> Stack {
 
         Ok(())
     }
+
+    /// Report a descriptive error when `new` could not be granted from `derived_from`.
+    fn grant_error(
+        &self,
+        derived_from: SbTag,
+        new: Item,
+        alloc_id: AllocId,
+        alloc_range: AllocRange,
+        error_offset: Size,
+    ) -> InterpError<'static> {
+        let action = format!(
+            "trying to reborrow {:?}+{:#x} with {:?} permission",
+            alloc_id,
+            error_offset.bytes(),
+            new.perm,
+        );
+        let help = Some(format!(
+            "this is a reborrow of {:?}[{:#x}..{:#x}]",
+            alloc_id,
+            alloc_range.start.bytes(),
+            alloc_range.end().bytes()
+        ));
+        if self
+            .borrows
+            .iter()
+            .any(|item| item.tag == derived_from && item.perm != Permission::Disabled)
+        {
+            err_sb_ub(
+                format!(
+                    "{}, but parent tag {:?} only grants SharedReadOnly permission for this memory",
+                    action, derived_from
+                ),
+                help,
+            )
+        } else {
+            err_sb_ub(
+                format!(
+                    "{}, but parent tag {:?} does not exist in the borrow stack for this memory",
+                    action, derived_from,
+                ),
+                help,
+            )
+        }
+    }
+
+    /// Report a descriptive error when `access` is not permitted based on `tag`.
+    fn access_error(
```

Comment:
Looks like there is quite a bit of duplication between these two functions? Is it possible and useful to share some code here?

Reply:
Took a shot at some deduplication. It sort of looks better?